packages/tinybased/src/lib/queries/Query.spec.ts

```diff
@@ -0,0 +1,211 @@
+import { SchemaBuilder } from '../SchemaBuilder';
+import { TableBuilder } from '../TableBuilder';
+
+const sb = new SchemaBuilder()
+  .addTable(new TableBuilder('users').add('id', 'string').add('name', 'string'))
+  .addTable(
+    new TableBuilder('notes')
+      .add('id', 'string')
+      .add('userId', 'string')
+      .add('likes', 'number')
+      .addOptional('isDraft', 'boolean')
+  )
+  .addTable(new TableBuilder('tags').add('id', 'string').add('name', 'string'))
+  .addTable(
+    new TableBuilder('noteTags')
+      .add('noteId', 'string')
+      .add('tagId', 'string')
+      .keyBy(['noteId', 'tagId'])
+  )
+  .defineRelationship('userNotes', 'notes', 'users', 'userId')
+  .defineRelationship('noteTagNotes', 'noteTags', 'notes', 'noteId')
+  .defineRelationship('noteTagTags', 'noteTags', 'tags', 'tagId')
+  .defineHydrators({
+    users: async () => [
+      {
+        id: 'user1',
+        name: 'Jesse',
+      },
+      {
+        id: 'user2',
+        name: 'Christyn',
+      },
+    ],
+    notes: async () => [
+      {
+        id: 'note1',
+        userId: 'user1',
+        likes: 4,
+        isDraft: true,
+      },
+      {
+        id: 'note2',
+        userId: 'user1',
+        likes: 2,
+        isDraft: false,
+      },
+      {
+        id: 'note3',
+        userId: 'user2',
+        likes: 1,
+        isDraft: false,
+      },
+    ],
+    noteTags: async () => [
+      {
+        noteId: 'note1',
+        tagId: 'tag1',
+      },
+      {
+        noteId: 'note2',
+        tagId: 'tag2',
+      },
+      {
+        noteId: 'note3',
+        tagId: 'tag3',
+      },
+    ],
+    tags: async () => [
+      {
+        id: 'tag1',
+        name: 'entertainment',
+      },
+      {
+        id: 'tag2',
+        name: 'work',
+      },
+      {
+        id: 'tag3',
+        name: 'sports',
+      },
+    ],
+  });
+
+describe('QueryBuilder', () => {
+  it('works with raw queries', async () => {
+    const db = await sb.build();
+
+    db.queries.setQueryDefinition(
+      'manyToMany',
+      'noteTags',
+      ({ where, join, select }) => {
+        join('notes', 'noteId');
+        join('tags', 'tagId');
+        where('notes', 'userId', 'user1');
+        select('tags', 'name');
+      }
+    );
+
+    const result = db.queries.getResultTable('manyToMany');
+    const tagNames = Object.values(result).map((v) => v['name']);
+    expect(tagNames).toEqual(['entertainment', 'work']);
+  });
+
+  it('supports typesafe joins', async () => {
+    const db = await sb.build();
+
+    const qb = db
+      .query('noteTags')
+      .join('noteTagNotes')
+      .join('noteTagTags')
+      .whereFrom('notes', 'userId', 'user1')
```

Comment:
When I was designing the API, GitHub copilot suggested `*From` for others but for where it suggested `whereIn`. Not sure there is any meaningful difference between `whereFrom` and `whereIn`.

Reply:
In normal SQL, `whereIn` generally deals with searching inside an array/collection. For right now I prefer to be explicit which select/where come from which joined tables